File: ProGO/backend/ml/python/performance_predictor.py
"""
Performance Prediction using TensorFlow/Keras
Advanced ML model with feature engineering for student performance prediction
"""

import json
import numpy as np
import pandas as pd
from tensorflow import keras
from tensorflow.keras import layers, callbacks
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
import joblib
import os
import sys
sys.path.append(os.path.dirname(__file__))
from utils.feature_engineering import PerformanceFeatureEngineer, create_synthetic_training_data

class PerformancePredictor:
    """Advanced Performance Prediction with Neural Network"""
    
    def __init__(self, 
                 model_path='models/performance_predictor.h5',
                 scaler_path='models/performance_scaler.pkl'):
        self.model_path = model_path
        self.scaler_path = scaler_path
        self.model = None
        self.scaler = StandardScaler()
        self.feature_engineer = PerformanceFeatureEngineer()
        self.feature_names = []
        
    def build_model(self, input_dim: int) -> keras.Model:
        """
        Build neural network model
        
        Args:
            input_dim: Number of input features
            
        Returns:
            Compiled Keras model
        """
        model = keras.Sequential([
            layers.Input(shape=(input_dim,)),
            layers.Dense(128, activation='relu'),
            layers.Dropout(0.3),
            layers.BatchNormalization(),
            layers.Dense(64, activation='relu'),
            layers.Dropout(0.2),
            layers.BatchNormalization(),
            layers.Dense(32, activation='relu'),
            layers.Dropout(0.2),
            layers.Dense(16, activation='relu'),
            layers.Dense(1, activation='linear')  # Regression output
        ])
        
        model.compile(
            optimizer=keras.optimizers.Adam(learning_rate=0.001),
            loss='mse',
            metrics=['mae', 'mse']
        )
        
        return model
    
    def train(self, use_synthetic: bool = True, n_samples: int = 2000):
        """
        Train the performance predictor
        
        Args:
            use_synthetic: Whether to use synthetic training data
            n_samples: Number of synthetic samples to generate
        """
        print(f"Generating {n_samples} training samples...")
        
        # Generate synthetic training data
        raw_data, targets = create_synthetic_training_data(n_samples)
        
        # Engineer features
        print("Engineering features...")
        engineered_features = []
        for idx in range(len(raw_data)):
            row_dict = raw_data.iloc[idx].to_dict()
            features = self.feature_engineer.create_features(row_dict)
            engineered_features.append(features[0])
        
        X = np.array(engineered_features)
        y = targets
        
        self.feature_names = self.feature_engineer.feature_names
        print(f"Created {X.shape[1]} features: {', '.join(self.feature_names)}")
        
        # Split data
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42
        )
        
        # Scale features
        print("Scaling features...")
        X_train_scaled = self.scaler.fit_transform(X_train)
        X_test_scaled = self.scaler.transform(X_test)
        
        # Build model
        print("Building neural network...")
        self.model = self.build_model(X_train_scaled.shape[1])
        
        print(f"\nModel Architecture:")
        self.model.summary()
        
        # Callbacks
        early_stop = callbacks.EarlyStopping(
            monitor='val_loss',
            patience=15,
            restore_best_weights=True,
            verbose=1
        )
        
        reduce_lr = callbacks.ReduceLROnPlateau(
            monitor='val_loss',
            factor=0.5,
            patience=5,
            min_lr=0.00001,
            verbose=1
        )
        
        # Train
        print("\nTraining model...")
        history = self.model.fit(
            X_train_scaled, y_train,
            validation_split=0.2,
            epochs=100,
            batch_size=32,
            callbacks=[early_stop, reduce_lr],
            verbose=1
        )
        
        # Evaluate
        print("\nEvaluating model...")
        train_loss, train_mae, train_mse = self.model.evaluate(X_train_scaled, y_train, verbose=0)
        test_loss, test_mae, test_mse = self.model.evaluate(X_test_scaled, y_test, verbose=0)
        
        # Predictions for metrics
        y_train_pred = self.model.predict(X_train_scaled, verbose=0).flatten()
        y_test_pred = self.model.predict(X_test_scaled, verbose=0).flatten()
        
        train_r2 = r2_score(y_train, y_train_pred)
        test_r2 = r2_score(y_test, y_test_pred)
        
        print(f"\n{'='*60}")
        print(f"Training Results:")
        print(f"{'='*60}")
        print(f"Train MAE: {train_mae:.4f} | Train MSE: {train_mse:.4f} | Train R²: {train_r2:.4f}")
        print(f"Test  MAE: {test_mae:.4f} | Test  MSE: {test_mse:.4f} | Test  R²: {test_r2:.4f}")
        print(f"{'='*60}")
        
        # Save model
        self.save()
        
        return {
            'train_mae': float(train_mae),
            'train_mse': float(train_mse),
            'train_r2': float(train_r2),
            'test_mae': float(test_mae),
            'test_mse': float(test_mse),
            'test_r2': float(test_r2),
            'n_samples': n_samples,
            'n_features': X.shape[1],
            'feature_names': self.feature_names,
            'epochs_trained': len(history.history['loss'])
        }
    
    def predict(self, student_data: dict) -> dict:
        """
        Predict student performance
        
        Args:
            student_data: Dictionary containing student metrics
            
        Returns:
            Prediction results with insights
        """
        if self.model is None:
            self.load()
        
        # Engineer features
        features = self.feature_engineer.create_features(student_data)
        
        # Scale
        features_scaled = self.scaler.transform(features)
        
        # Predict
        prediction = self.model.predict(features_scaled, verbose=0)[0][0]
        predicted_cgpa = float(np.clip(prediction, 0, 10))
        
        # Generate insights
        insights = self._generate_insights(student_data, features[0], predicted_cgpa)
        
        return {
            'predicted_cgpa': predicted_cgpa,
            'current_cgpa': student_data.get('cgpa', 0),
            'improvement_potential': float(predicted_cgpa - student_data.get('cgpa', 0)),
            'confidence': self._calculate_confidence(student_data),
            'insights': insights,
            'recommendations': self._generate_recommendations(student_data, features[0])
        }
    
    def _calculate_confidence(self, data: dict) -> float:
        """Calculate prediction confidence based on data quality"""
        confidence = 1.0
        
        # Reduce confidence for missing or unrealistic data
        if data.get('attendance', 0) == 0:
            confidence *= 0.7
        if data.get('studyHours', 0) == 0:
            confidence *= 0.8
        if data.get('cgpa', 0) == 0:
            confidence *= 0.6
            
        return float(confidence)
    
    def _generate_insights(self, data: dict, features: np.ndarray, predicted_cgpa: float) -> list:
        """Generate actionable insights"""
        insights = []
        
        # Attendance insight
        attendance = data.get('attendance', 0)
        if attendance < 75:
            insights.append({
                'type': 'warning',
                'category': 'attendance',
                'message': f'Low attendance ({attendance}%) is a major concern. Aim for 75%+ to avoid detention.',
                'priority': 'high'
            })
        elif attendance >= 90:
            insights.append({
                'type': 'positive',
                'category': 'attendance',
                'message': f'Excellent attendance ({attendance}%)! Keep it up.',
                'priority': 'low'
            })
        
        # CGPA trend
        current_cgpa = data.get('cgpa', 0)
        improvement = predicted_cgpa - current_cgpa
        if improvement > 0.5:
            insights.append({
                'type': 'positive',
                'category': 'performance',
                'message': f'Strong improvement trajectory! Predicted CGPA: {predicted_cgpa:.2f} (up from {current_cgpa:.2f})',
                'priority': 'medium'
            })
        elif improvement < -0.3:
            insights.append({
                'type': 'warning',
                'category': 'performance',
                'message': f'Performance may decline to {predicted_cgpa:.2f}. Take corrective action now.',
                'priority': 'high'
            })
        
        # Backlog warning
        backlogs = data.get('backlogs', 0)
        if backlogs > 0:
            insights.append({
                'type': 'warning',
                'category': 'backlog',
                'message': f'You have {backlogs} backlog(s). Clear them as soon as possible.',
                'priority': 'high'
            })
        
        # Study hours
        study_hours = data.get('studyHours', 0)
        if study_hours < 3:
            insights.append({
                'type': 'warning',
                'category': 'study_habits',
                'message': f'Low study hours ({study_hours}h/day). Recommend 4-6 hours for optimal results.',
                'priority': 'medium'
            })
        
        # Academic consistency (feature index 10)
        if len(features) > 10:
            consistency = features[10]
            if consistency < 0.7:
                insights.append({
                    'type': 'warning',
                    'category': 'consistency',
                    'message': 'Inconsistent performance across subjects. Focus on weaker areas.',
                    'priority': 'medium'
                })
        
        return insights
    
    def _generate_recommendations(self, data: dict, features: np.ndarray) -> list:
        """Generate personalized recommendations"""
        recommendations = []
        
        # Attendance recommendations
        attendance = data.get('attendance', 0)
        if attendance < 75:
            recommendations.append({
                'action': 'Attend all upcoming classes regularly',
                'impact': 'high',
                'category': 'attendance'
            })
        
        # Study recommendations
        study_hours = data.get('studyHours', 0)
        if study_hours < 4:
            target_hours = 5
            recommendations.append({
                'action': f'Increase study time from {study_hours}h to {target_hours}h per day',
                'impact': 'high',
                'category': 'study_habits'
            })
        
        # Backlog clearing
        backlogs = data.get('backlogs', 0)
        if backlogs > 0:
            recommendations.append({
                'action': f'Prioritize clearing {backlogs} backlog subject(s) this semester',
                'impact': 'critical',
                'category': 'backlog'
            })
        
        # Score improvement
        theory_score = data.get('theoryScore', 0)
        practical_score = data.get('practicalScore', 0)
        
        if theory_score < 60 or practical_score < 60:
            recommendations.append({
                'action': 'Focus on fundamentals - revise weak topics regularly',
                'impact': 'high',
                'category': 'academics'
            })
        
        # Balanced performance
        if abs(theory_score - practical_score) > 20:
            recommendations.append({
                'action': 'Balance theory and practical preparation',
                'impact': 'medium',
                'category': 'academics'
            })
        
        # Engagement
        if len(features) > 9:
            engagement = features[9]  # engagement_score
            if engagement < 0.5:
                recommendations.append({
                    'action': 'Increase classroom participation and self-study time',
                    'impact': 'high',
                    'category': 'engagement'
                })
        
        return recommendations
    
    def save(self):
        """Save model and scaler"""
        os.makedirs(os.path.dirname(self.model_path), exist_ok=True)
        
        # Save Keras model
        self.model.save(self.model_path)
        print(f"Model saved to {self.model_path}")
        
        # Save scaler and feature engineer
        joblib.dump({
            'scaler': self.scaler,
            'feature_engineer': self.feature_engineer,
            'feature_names': self.feature_names
        }, self.scaler_path)
        print(f"Scaler saved to {self.scaler_path}")
    
    def load(self):
        """Load model and scaler"""
        if not os.path.exists(self.model_path):
            raise FileNotFoundError(f"Model not found at {self.model_path}. Please train first.")
        
        # Load Keras model
        self.model = keras.models.load_model(self.model_path)
        print(f"Model loaded from {self.model_path}")
        
        # Load scaler and feature engineer
        data = joblib.load(self.scaler_path)
        self.scaler = data['scaler']
        self.feature_engineer = data['feature_engineer']
        self.feature_names = data['feature_names']


# CLI Interface
if __name__ == '__main__':
    predictor = PerformancePredictor()
    
    if len(sys.argv) > 1:
        command = sys.argv[1]
        
        if command == 'train':
            # Train the model
            n_samples = int(sys.argv[2]) if len(sys.argv) > 2 else 2000
            stats = predictor.train(n_samples=n_samples)
            print(json.dumps(stats, indent=2))
            
        elif command == 'predict':
            # Predict performance for student data (JSON input)
            if len(sys.argv) < 3:
                print("Usage: python performance_predictor.py predict <json_data>")
                sys.exit(1)
            
            student_data = json.loads(sys.argv[2])
            result = predictor.predict(student_data)
            print(json.dumps(result, indent=2))
            
        elif command == 'test':
            # Test with sample data
            test_cases = [
                {
                    'name': 'Excellent Student',
                    'data': {
                        'attendance': 95,
                        'cgpa': 8.5,
                        'semester': 4,
                        'backlogs': 0,
                        'studyHours': 6,
                        'theoryScore': 85,
                        'practicalScore': 88,
                        'assignmentScore': 90
                    }
                },
                {
                    'name': 'Average Student',
                    'data': {
                        'attendance': 75,
                        'cgpa': 6.5,
                        'semester': 5,
                        'backlogs': 1,
                        'studyHours': 4,
                        'theoryScore': 65,
                        'practicalScore': 70,
                        'assignmentScore': 68
                    }
                },
                {
                    'name': 'Struggling Student',
                    'data': {
                        'attendance': 60,
                        'cgpa': 5.2,
                        'semester': 3,
                        'backlogs': 3,
                        'studyHours': 2,
                        'theoryScore': 50,
                        'practicalScore': 55,
                        'assignmentScore': 52
                    }
                }
            ]
            
            results = []
            for case in test_cases:
                print(f"\n{'='*60}")
                print(f"Testing: {case['name']}")
                print(f"{'='*60}")
                result = predictor.predict(case['data'])
                results.append({
                    'name': case['name'],
                    'prediction': result
                })
                print(f"Current CGPA: {result['current_cgpa']:.2f}")
                print(f"Predicted CGPA: {result['predicted_cgpa']:.2f}")
                print(f"Improvement: {result['improvement_potential']:.2f}")
                print(f"Confidence: {result['confidence']:.2f}")
                print(f"\nInsights:")
                for insight in result['insights']:
                    print(f"  [{insight['priority']}] {insight['message']}")
                print(f"\nRecommendations:")
                for rec in result['recommendations']:
                    print(f"  [{rec['impact']}] {rec['action']}")
            
            print(f"\n{'='*60}")
            print(json.dumps(results, indent=2))
    else:
        print("Usage:")
        print("  python performance_predictor.py train [n_samples]")
        print("  python performance_predictor.py predict '<json_data>'")
        print("  python performance_predictor.py test")

File: ProGO/backend/ml/performancePredictor.js
const brain = require('brain.js');
const fs = require('fs');
const path = require('path');

class PerformancePredictor {
  constructor() {
    this.network = new brain.NeuralNetwork({
      hiddenLayers: [64, 32, 16],
      activation: 'sigmoid',
      learningRate: 0.01
    });
    this.modelPath = path.join(__dirname, 'models', 'performance-model.json');
    this.isInitialized = false;
  }

  // Initialize or load model
  async initialize() {
    try {
      if (fs.existsSync(this.modelPath)) {
        console.log('📊 Loading Performance Prediction Model...');
        await this.loadModel();
      } else {
        console.log('📊 Performance Prediction Model ready for training...');
      }
      this.isInitialized = true;
      console.log('✓ Performance Predictor initialized');
    } catch (error) {
      console.error('✗ Error initializing Performance Predictor:', error);
      throw error;
    }
  }

  // Generate sample training data for initial model training
  generateSampleTrainingData(sampleCount = 1000) {
    const trainingData = [];

    for (let i = 0; i < sampleCount; i++) {
      // Generate random but realistic student data
      const attendance = 50 + Math.random() * 50; // 50-100%
      const previousCGPA = 5 + Math.random() * 5; // 5-10
      const assignmentScore = 40 + Math.random() * 60; // 40-100
      const midtermScore = 40 + Math.random() * 60; // 40-100
      const participationScore = Math.random() * 10; // 0-10
      const backlogs = Math.floor(Math.random() * 5); // 0-4
      const studyHours = 5 + Math.random() * 20; // 5-25 hours/week
      const semester = Math.floor(Math.random() * 8) + 1; // 1-8

      // Calculate next semester CGPA with realistic formula
      let nextSemesterCGPA = 
        (attendance * 0.15 / 10) +
        (previousCGPA * 0.40) +
        (assignmentScore * 0.15 / 10) +
        (midtermScore * 0.15 / 10) +
        (participationScore * 0.05) -
        (backlogs * 0.3) +
        (studyHours * 0.05);

      // Add some randomness
      nextSemesterCGPA += (Math.random() - 0.5) * 0.5;

      // Normalize to 0-10 range
      nextSemesterCGPA = Math.max(0, Math.min(10, nextSemesterCGPA));

      trainingData.push({
        attendance,
        previousCGPA,
        assignmentScore,
        midtermScore,
        participationScore,
        backlogs,
        studyHours,
        semester,
        nextSemesterCGPA
      });
    }

    return trainingData;
  }

  // Load training data from CSV dataset
  loadTrainingDataFromCSV(csvPath) {
    if (!fs.existsSync(csvPath)) {
      throw new Error(`Dataset not found at ${csvPath}`);
    }

    const raw = fs.readFileSync(csvPath, 'utf8').trim();
    if (!raw) {
      throw new Error('Dataset CSV is empty');
    }

    const lines = raw.split(/\r?\n/);
    const headers = this.parseCsvLine(lines[0]);
    const rows = lines.slice(1).map((line) => this.parseCsvLine(line));

    return rows
      .filter((cols) => cols.length === headers.length)
      .map((cols) => {
        const row = {};
        headers.forEach((header, idx) => {
          row[header] = cols[idx];
        });

        const attendance = parseFloat(row.attendance_percentage || 0);
        const previousCGPA = parseFloat(row.previous_cgpa || 0);
        const assignmentsCompleted = parseFloat(row.assignments_completed || 0);
        const extracurricular = parseFloat(row.extracurricular_activities || 0);
        const semester = parseInt(row.semester || 1, 10);
        const studyHoursPerDay = parseFloat(row.study_hours_per_day || 0);
        const hasBacklogs = String(row['Do I have any backlogs?'] || '').toLowerCase() === 'yes';

        const studyHours = Math.min(studyHoursPerDay * 4, 25);
        const assignmentScore = Math.max(0, Math.min(100, assignmentsCompleted));
        const participationScore = Math.max(0, Math.min(10, extracurricular * 1.6));
        const backlogs = hasBacklogs ? Math.max(1, Math.round((10 - previousCGPA) / 2)) : 0;
        const midtermScore = Math.max(
          40,
          Math.min(100, previousCGPA * 10 + (attendance - 75) * 0.35 + (Math.random() - 0.5) * 8)
        );

        let nextSemesterCGPA =
          previousCGPA * 0.6 +
          (attendance / 100) * 1.3 +
          (assignmentScore / 100) * 0.9 +
          (midtermScore / 100) * 1.0 +
          (participationScore / 10) * 0.4 -
          backlogs * 0.35 +
          (studyHours / 25) * 0.6 +
          (Math.random() - 0.5) * 0.25;

        if (row.performance_category === 'Excellent') nextSemesterCGPA += 0.25;
        if (row.performance_category === 'Needs Improvement') nextSemesterCGPA -= 0.25;

        nextSemesterCGPA = Math.max(0, Math.min(10, nextSemesterCGPA));

        return {
          attendance,
          previousCGPA,
          assignmentScore,
          midtermScore,
          participationScore,
          backlogs,
          studyHours,
          semester,
          nextSemesterCGPA
        };
      });
  }

  parseCsvLine(line) {
    const values = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
      const ch = line[i];

      if (ch === '"') {
        if (inQuotes && line[i + 1] === '"') {
          current += '"';
          i++;
        } else {
          inQuotes = !inQuotes;
        }
        continue;
      }

      if (ch === ',' && !inQuotes) {
        values.push(current);
        current = '';
      } else {
        current += ch;
      }
    }

    values.push(current);
    return values;
  }

  // Train the model with historical data
  async train(trainingData) {
    try {
      console.log('🎓 Training Performance Prediction Model...');

      // Prepare training data for Brain.js
      const brainTrainingData = trainingData.map(d => ({
        input: [
          d.attendance / 100,           // Normalized attendance
          d.previousCGPA / 10,           // Normalized CGPA
          d.assignmentScore / 100,       // Normalized assignment score
          d.midtermScore / 100,          // Normalized midterm score
          d.participationScore / 10,     // Normalized participation
          d.backlogs / 10,               // Normalized backlogs count
          d.studyHours / 25,             // Normalized study hours (assuming max 25)
          d.semester / 8                 // Normalized semester number
        ],
        output: [d.nextSemesterCGPA / 10] // Normalized target
      }));

      console.log(`  📊 Training with ${brainTrainingData.length} samples...`);

      const startTime = Date.now();

      const stats = this.network.train(brainTrainingData, {
        iterations: 5000,
        errorThresh: 0.005,
        log: true,
        logPeriod: 500,
        learningRate: 0.01,
        callback: (stats) => {
          if (stats.iterations % 500 === 0) {
            console.log(`  ⏳ Iteration ${stats.iterations}: error = ${stats.error.toFixed(6)}`);
          }
        }
      });

      const endTime = Date.now();
      const duration = ((endTime - startTime) / 1000).toFixed(2);

      console.log(`✓ Model trained successfully in ${duration}s. Final error: ${stats.error.toFixed(6)}`);
      
      // Save the trained model
      await this.saveModel();
      
      return stats;
    } catch (error) {
      console.error('✗ Error training model:', error);
      throw error;
    }
  }

  // Predict next semester CGPA
  async predict(studentData) {
    try {
      if (!this.isInitialized) {
        await this.initialize();
      }

      const input = [
        studentData.attendance / 100,
        studentData.previousCGPA / 10,
        studentData.assignmentScore / 100,
        studentData.midtermScore / 100,
        studentData.participationScore / 10,
        studentData.backlogs / 10,
        studentData.studyHours / 25,
        studentData.semester / 8
      ];

      const output = this.network.run(input);
      const predictedCGPA = output[0] * 10;

      return {
        predictedCGPA: parseFloat(predictedCGPA.toFixed(2)),
        confidence: this.calculatePredictionConfidence(studentData),
        factors: this.analyzeContributingFactors(studentData)
      };
    } catch (error) {
      console.error('Error predicting performance:', error);
      throw error;
    }
  }

  // Calculate prediction confidence based on data quality
  calculatePredictionConfidence(studentData) {
    let confidence = 1.0;

    // Reduce confidence if we have incomplete data
    if (!studentData.attendance) confidence *= 0.9;
    if (!studentData.previousCGPA) confidence *= 0.9;
    if (!studentData.assignmentScore) confidence *= 0.95;
    if (!studentData.midtermScore) confidence *= 0.95;

    // Reduce confidence for edge cases
    if (studentData.attendance < 60) confidence *= 0.85;
    if (studentData.backlogs > 3) confidence *= 0.85;

    return parseFloat(confidence.toFixed(2));
  }

  // Analyze which factors contribute most to performance
  analyzeContributingFactors(studentData) {
    const factors = [];

    // Attendance analysis
    if (studentData.attendance >= 85) {
      factors.push({ factor: 'Attendance', impact: 'positive', weight: 'high', value: `${studentData.attendance}%` });
    } else if (studentData.attendance >= 75) {
      factors.push({ factor: 'Attendance', impact: 'neutral', weight: 'medium', value: `${studentData.attendance}%` });
    } else {
      factors.push({ factor: 'Attendance', impact: 'negative', weight: 'high', value: `${studentData.attendance}%` });
    }

    // Previous CGPA analysis
    if (studentData.previousCGPA >= 8) {
      factors.push({ factor: 'Previous CGPA', impact: 'positive', weight: 'very high', value: studentData.previousCGPA });
    } else if (studentData.previousCGPA >= 6) {
      factors.push({ factor: 'Previous CGPA', impact: 'neutral', weight: 'high', value: studentData.previousCGPA });
    } else {
      factors.push({ factor: 'Previous CGPA', impact: 'negative', weight: 'very high', value: studentData.previousCGPA });
    }

    // Backlogs analysis
    if (studentData.backlogs === 0) {
      factors.push({ factor: 'Backlogs', impact: 'positive', weight: 'high', value: 0 });
    } else if (studentData.backlogs <= 2) {
      factors.push({ factor: 'Backlogs', impact: 'negative', weight: 'medium', value: studentData.backlogs });
    } else {
      factors.push({ factor: 'Backlogs', impact: 'negative', weight: 'high', value: studentData.backlogs });
    }

    // Study hours analysis
    if (studentData.studyHours >= 18) {
      factors.push({ factor: 'Study Hours', impact: 'positive', weight: 'medium', value: `${studentData.studyHours}/week` });
    } else if (studentData.studyHours >= 12) {
      factors.push({ factor: 'Study Hours', impact: 'neutral', weight: 'low', value: `${studentData.studyHours}/week` });
    } else {
      factors.push({ factor: 'Study Hours', impact: 'negative', weight: 'medium', value: `${studentData.studyHours}/week` });
    }

    return factors;
  }

  // Analyze performance and provide insights
  async analyzePerformance(studentData) {
    try {
      const prediction = await this.predict(studentData);
      
      const insights = {
        prediction: prediction.predictedCGPA,
        confidence: prediction.confidence,
        currentPerformance: this.categorizePerformance(studentData.previousCGPA),
        predictedPerformance: this.categorizePerformance(prediction.predictedCGPA),
        trend: this.analyzeTrend(studentData.previousCGPA, prediction.predictedCGPA),
        contributingFactors: prediction.factors,
        recommendations: this.generateRecommendations(studentData, prediction)
      };

      return insights;
    } catch (error) {
      console.error('Error analyzing performance:', error);
      throw error;
    }
  }

  // Categorize performance level
  categorizePerformance(cgpa) {
    if (cgpa >= 9) return 'Excellent';
    if (cgpa >= 8) return 'Very Good';
    if (cgpa >= 7) return 'Good';
    if (cgpa >= 6) return 'Satisfactory';
    if (cgpa >= 5) return 'Needs Improvement';
    return 'Critical';
  }

  // Analyze performance trend
  analyzeTrend(currentCGPA, predictedCGPA) {
    const diff = predictedCGPA - currentCGPA;
    
    if (diff >= 0.5) return { direction: 'improving', magnitude: 'significant', change: diff };
    if (diff >= 0.2) return { direction: 'improving', magnitude: 'moderate', change: diff };
    if (diff > -0.2) return { direction: 'stable', magnitude: 'minimal', change: diff };
    if (diff > -0.5) return { direction: 'declining', magnitude: 'moderate', change: diff };
    return { direction: 'declining', magnitude: 'significant', change: diff };
  }

  // Generate personalized recommendations
  generateRecommendations(studentData, prediction) {
    const recommendations = [];

    // Attendance recommendation
    if (studentData.attendance < 85) {
      recommendations.push({
        area: 'Attendance',
        priority: studentData.attendance < 75 ? 'critical' : 'high',
        message: `Increase attendance to at least 85%. Current: ${studentData.attendance}%`,
        potentialImpact: '+0.4 CGPA points'
      });
    }

    // Backlogs recommendation
    if (studentData.backlogs > 0) {
      recommendations.push({
        area: 'Backlogs',
        priority: 'critical',
        message: `Clear ${studentData.backlogs} pending backlog(s) immediately`,
        potentialImpact: '+0.5 CGPA points'
      });
    }

    // Study hours recommendation
    if (studentData.studyHours < 15) {
      recommendations.push({
        area: 'Study Time',
        priority: 'medium',
        message: `Increase study hours to at least 15-20 hours per week. Current: ${studentData.studyHours}`,
        potentialImpact: '+0.3 CGPA points'
      });
    }

    // Assignment scores recommendation
    if (studentData.assignmentScore < 80) {
      recommendations.push({
        area: 'Assignments',
        priority: 'medium',
        message: 'Focus on completing assignments with better quality',
        potentialImpact: '+0.2 CGPA points'
      });
    }

    // Midterm scores recommendation
    if (studentData.midtermScore < 75) {
      recommendations.push({
        area: 'Exam Preparation',
        priority: 'high',
        message: 'Improve exam preparation and revision strategies',
        potentialImpact: '+0.3 CGPA points'
      });
    }

    return recommendations.sort((a, b) => {
      const priorityOrder = { critical: 0, high: 1, medium: 2, low: 3 };
      return priorityOrder[a.priority] - priorityOrder[b.priority];
    });
  }

  // Train from database records
  async trainFromDatabase(studentRecords) {
    try {
      console.log('📚 Training from database records...');
      
      const trainingData = studentRecords.map(record => ({
        attendance: record.attendance || 75,
        previousCGPA: record.previousCGPA || 7,
        assignmentScore: record.assignmentScore || 75,
        midtermScore: record.midtermScore || 75,
        participationScore: record.participationScore || 5,
        backlogs: record.backlogs || 0,
        studyHours: record.studyHours || 15,
        semester: record.semester || 1,
        nextSemesterCGPA: record.nextSemesterCGPA || record.previousCGPA
      }));

      await this.train(trainingData);
      console.log('✓ Model trained from database successfully');
    } catch (error) {
      console.error('Error training from database:', error);
      throw error;
    }
  }

  // Save trained model to file
  async saveModel() {
    try {
      const modelDir = path.join(__dirname, 'models');
      if (!fs.existsSync(modelDir)) {
        fs.mkdirSync(modelDir, { recursive: true });
      }

      const model = this.network.toJSON();
      fs.writeFileSync(this.modelPath, JSON.stringify(model));
      console.log('✓ Performance model saved to:', this.modelPath);
    } catch (error) {
      console.error('✗ Error saving model:', error);
    }
  }

  // Load trained model from file
  async loadModel() {
    try {
      if (fs.existsSync(this.modelPath)) {
        const model = JSON.parse(fs.readFileSync(this.modelPath, 'utf8'));
        this.network.fromJSON(model);
        console.log('✓ Performance Prediction model loaded');
        return true;
      } else {
        console.log('⚠ No trained model found');
        return false;
      }
    } catch (error) {
      console.error('✗ Error loading model:', error);
      return false;
    }
  }
}

module.exports = new PerformancePredictor();
